src/lib/passMetadataManager.ts

```diff
@@ -0,0 +1,302 @@
+import * as VIAM from "@viamrobotics/sdk";
+import { PassNote, PassDiagnosis } from '../types';
+
+const NOTE_PREFIX = 'note-';
+const DIAGNOSIS_PREFIX = 'diagnosis-';
+
+export class PassMetadataManager {
+  private viamClient: VIAM.ViamClient;
+  private machineId: string;
+  private cachedNotes: Map<string, PassNote> | null = null;
+  private cachedDiagnoses: Map<string, PassDiagnosis> | null = null;
+
+  constructor(viamClient: VIAM.ViamClient, machineId: string) {
+    this.viamClient = viamClient;
+    this.machineId = machineId;
+  }
+
+  /**
+   * Generic helper to parse metadata items by prefix
+   * Reduces duplication between notes and diagnoses parsing
+   */
+  private parseMetadataByPrefix<T>(
+    metadata: Record<string, unknown>,
+    prefix: string,
+    entityName: string
+  ): Map<string, T> {
+    const result = new Map<string, T>();
+
+    Object.keys(metadata).forEach(key => {
+      if (key.startsWith(prefix)) {
+        const passId = key.substring(prefix.length);
+        try {
+          const data = JSON.parse(metadata[key] as string);
+          result.set(passId, data as T);
+        } catch (e) {
+          console.warn(`Failed to parse ${entityName} for pass ${passId}:`, e);
+        }
+      }
+    });
+
+    return result;
+  }
+
+  /**
+   * Generic helper to update metadata with merge
+   * Fetches current metadata, applies changes, and saves back
+   */
+  private async updateMetadata(
+    updateFn: (metadata: Record<string, unknown>) => void
+  ): Promise<void> {
+    const currentMetadata = await this.viamClient.appClient.getRobotMetadata(this.machineId);
+    updateFn(currentMetadata);
+    await this.viamClient.appClient.updateRobotMetadata(this.machineId, currentMetadata);
+  }
+
+  /**
+   * Get all notes from metadata - stored as flat key-value pairs
+   * Each note is stored as: "note-{passId}": "json-stringified-PassNote"
+   */
+  private async getNotesMetadata(): Promise<Map<string, PassNote>> {
+    if (this.cachedNotes !== null) {
+      return this.cachedNotes;
+    }
+
+    const metadata = await this.viamClient.appClient.getRobotMetadata(this.machineId);
+    this.cachedNotes = this.parseMetadataByPrefix<PassNote>(metadata, NOTE_PREFIX, 'note');
+
+    return this.cachedNotes;
+  }
+
+  /**
+   * Get all diagnoses from metadata - stored as flat key-value pairs
+   * Each diagnosis is stored as: "diagnosis-{passId}": "json-stringified-PassDiagnosis"
+   */
+  private async getDiagnosesMetadata(): Promise<Map<string, PassDiagnosis>> {
+    if (this.cachedDiagnoses !== null) {
+      return this.cachedDiagnoses;
+    }
+
+    const metadata = await this.viamClient.appClient.getRobotMetadata(this.machineId);
```

Comment:
Is this returning all metadata from all passes? Are you worried about latency/storage here, slowing down / overwhelming the operation respectively?